File: Fraction pairs with sum 1 - GFG/fraction-pairs-with-sum-1.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;


class IntArray
{
    public static int[] input(BufferedReader br, int n) throws IOException
    {
        String[] s = br.readLine().trim().split(" ");
        int[] a = new int[n];
        for(int i = 0; i < n; i++)
            a[i] = Integer.parseInt(s[i]);

        return a;
    }

    public static void print(int[] a)
    {
        for(int e : a)
            System.out.print(e + " ");
        System.out.println();
    }

    public static void print(ArrayList<Integer> a)
    {
        for(int e : a)
            System.out.print(e + " ");
        System.out.println();
    }
}

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while(t-- > 0){
            
            int n;
            n = Integer.parseInt(br.readLine());
            
            
            int[] numerator = IntArray.input(br, n);
            
            
            int[] denominator = IntArray.input(br, n);
            
            Solution obj = new Solution();
            int res = obj.countFractions(n, numerator, denominator);
            
            System.out.println(res);
            
        }
    }
}

// } Driver Code Ends



// class Solution {
//     public static int countFractions(int n, int[] numerator, int[] denominator) {
//         // code here
//         int count = 0;
        
//         for (int i = 0; i < n; i++) {
//             double n1 = (double) numerator[i] / denominator[i];
//             for (int j = i + 1; j < n; j++) {
//                 double n2 = (double) numerator[j] / denominator[j];
//                 if (n1 + n2 == 1) {
//                     count++;
//                 }
//             }
//         }
//         return count;
//     }
// }
        class Solution {
    public static int countFractions(int n, int[] num, int[] denom) {
       int ans=0;
        HashMap<Integer,HashMap<Integer,Integer>> map=new HashMap<>();
        for(int i=0;i<n;i++)
        {
            int x=num[i],y=denom[i];
            int g=gcd(x,y);
            x/=g;  y/=g;
           
           int n_x = y-x; int n_y=y;
           //checking if 1-x/y is present in map - if yes -incerement the cnt;
            if(map.containsKey(n_x) && map.get(n_x).containsKey(y)){
                ans+=map.get(n_x).get(y);
            }
            //checking if some fractions have same num - push the denom in value(itself map)
            if(map.containsKey(x)){
                HashMap<Integer,Integer> hm =map.get(x);
                hm.put(y,hm.getOrDefault(y,0)+1);
                
            }
            //new fraction coming up.
            else{
                HashMap<Integer,Integer> hm=new HashMap<>();
                hm.put(y,1);
                map.put(x,hm);
            }
        }
        return ans;
    }
     static int gcd(int a,int b)
    {
        if(b==0)return a;
        return gcd(b,a%b);
    }
}

File: Fraction pairs with sum 1 - GFG/fraction-pairs-with-sum-1.cpp
//{ Driver Code Starts
//Initial Template for C++
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++

class Solution
{
    public:
    int countFractions(int n, int num[], int denom[])
    {
         int ans = 0;
        unordered_map<int, map<int, int>> mp;
        for (int i = 0; i < n; i++) {
            int x = num[i], y = denom[i];
            int g = gcd(x, y);
            x /= g; y /= g;
            
            int n_x = y - x, n_y = y;
            
            if (mp.find(n_x) != mp.end() && mp[n_x].find(y) != mp[n_x].end()) {
                ans += mp[n_x][y];
            }
            
            if (mp.find(x) != mp.end()) {
                map<int, int>& hm = mp[x];
                hm[y] = hm.find(y) != hm.end() ? hm[y] + 1 : 1;
            } else {
                map<int, int> hm;
                hm[y] = 1;
                mp[x] = hm;
            }
        }
        return ans;
    }

     int gcd(int a, int b) {
        if (b == 0) return a;
        return gcd(b, a % b);
    } 
    
};

//{ Driver Code Starts.
int main()
{
    int t;
    cin>>t;
    while(t--)
    {
        int n;
        cin>>n;
        int numerator[n],denominator[n];
        for(int i=0;i<n;i++)
            cin>>numerator[i];
        for(int i=0;i<n;i++)
            cin>>denominator[i];
        Solution ob;
        int ans=ob.countFractions(n,numerator,denominator);
        cout<<ans<<endl;
    }
}
// } Driver Code Ends